shell/browser/native_window_mac.mm

```diff
@@ -792,7 +793,24 @@ void ViewDidMoveToSuperview(NSView* self, SEL _cmd) {
   return [window_ isMiniaturized];
 }
 
+void NativeWindowMac::HandlePendingFullscreenTransitions() {
+  if (pending_transitions_.empty())
+    return;
+
+  bool next_transition = pending_transitions_.pop();
+  SetFullScreen(next_transition);
+}
+
 void NativeWindowMac::SetFullScreen(bool fullscreen) {
+  // [NSWindow -toggleFullScreen] is an asynchronous operation, which means
+  // that it's possible to call it while a fullscreen transition is currently
+  // in process. This can create weird behavior (incl. phantom windows),
+  // so we want to schedule a transition for when the current one has completed.
+  if (fullscreen_transition_state() != FullScreenTransitionState::NONE) {
```

Comment:
We can check the tail of `pending_transitions_` to see if we actually need to push a new transition.

If `pending_transitions_.back() == fullscreen` then the call can be dropped:

```js
setFullScreen(true);
setFullScreen(true);  // can be dropped
```

if  `pending_transitions_.back() != fullscreen` then the call can be dropped:

```js
setFullScreen(false);
setFullScreen(true);  // pending
setFullScreen(false);  // can cancel the pending call
```

We probably does not need a queue at all since with this optimization the queue's length won't be larger than 1. Something like `base::Optional<bool> pending_transition_` might be enough to do the work.

```c++
if (pending_transition_) {
  if (pending_transition_.get() == fullscreen) {
    // ignored as they are the same call.
  } else {
    pending_transition_.reset();  // current call cancels the pending transition.
  }
  return;
}
```

Reply:
I see the goals for this but i also think devs may want to actually complete the transitions - optimizing away the visual behavior is perhaps better but also less intuitive and may lead to confused devs wondering why the transitions seem never to have occurred 🤔

Reply:
I'm good not optimizing the transitions, but we still need to do something about duplicate calls. 

Since the duplicate calls won't actually do anything, the related events will not be emitted and the `HandlePendingFullscreenTransitions` method won't be called. So if there are still pending transitions they won't be processed after duplicate calls.


```js
setFullScreen(true);
setFullScreen(true);  // does not trigger events
setFullScreen(true);  // will be left in pending transitions unprocessed
```